rfcs/01780_policy_delegation.md

```diff
@@ -0,0 +1,308 @@
+manifest into the corresponding list of Wasm hashes.
+
+When an Oak Application creates a new Wasm Node, the Oak Runtime checks the hash
+of the Wasm module to instantiate, and if it correspond to one or more of the
+manifests, then it grants the newly created node the corresponding privileges.
+
+For instance, let us assume the following manifests:
+
+- `https://google.com/oak/manifests/count_people_from_pixels.json`:
+
+  ```json
+  {
+    "trusted": [
+      {
+        "type": "wasm_module",
+        "value": "0000000000000000000000000000000000000000"
+      },
+      {
+        "type": "wasm_module",
+        "value": "1111111111111111111111111111111111111111"
+      }
+    ]
+  }
+  ```
+
+- `https://other-company.com/oak/manifests/count_people_from_pixels.json`:
+
+  ```json
+  {
+    "trusted": [
+      {
+        "type": "wasm_module",
+        "value": "1111111111111111111111111111111111111111"
+      },
+      {
+        "type": "wasm_module",
+        "value": "2222222222222222222222222222222222222222"
+      }
+    ]
+  }
+  ```
+
+When an Oak Node with hash `1111111111111111111111111111111111111111` is
+instantiated, it receives the declassification privilege of
+"manifest:https://google.com/oak/manifests/count_people_from_pixels.json ∧
+manifest:https://other-company.com/oak/manifests/count_people_from_pixels.json ∧
+wasm_hash:1111111111111111111111111111111111111111".
+
+Note that in this case it receives both privileges (or rather, the conjunction
+between these privileges), plus the privilege for the corresponding Wasm hash
+label, which it would get anyways.
+
+In particular, the conjunctive privilege is quite useful in practice as a
+confidentiality label, because it may be used by the data owner to label data
+that is expected to be declassified by a Wasm module that is trusted by multiple
+parties at the same time, where each party independently publishes a separate
+manifest.
+
+## User-based delegation
+
+So far we have talked about companies being the publisher of such manifest, but
+it is possible for individual users to also publish manifests for specific use
+cases. A user may publish a manifest on a URL hosted on a server they control,
+or even as part of another website, e.g. on github.com. Of course, in the case
+of publishing on a different website, that website is the ultimate root of trust
+of any assertions.
+
+This allows delegation based on existing human-to-human network connections. For
+instance, a person may be an expert in a particular field, and independently
+verify properties of pieces of software and publish them in a manifest. Also it
+is possible for a person publishing a manifest to further delegate to another
+publisher (either company or person) by including the URL of the target manifest
+as a trusted object in their own manifest.
+
+For instance, a user may publish a manifest at the following URL:
+`https://github.com/tiziano88/delegation/count_users_from_pixels.json`, which is
```

Comment:
This seems problematic. The trusting user now has to trust the user publishing the manifest, github's staff, the repo's ACL configurations, the staff of the cloud provider hosting github, that there are no vulnerabilities in the github system that allows attackers to modify a file etc.

Reply:
Then maybe they should not use github, and set up their own web server (which the URL scheme allows, this solution is not tied to github in any way, it was just an example). Accessing code over github is already subject to all the issues you have described.

Reply:
An individual using their own web server seems even worse. The trust now also relies on their ability to correctly and securely configure their own web server and consistently maintain the configuraiton, security and patching. It also relies on them having a proper key management solution to securely manage the private key for their certificate.

Reply:
Then hopefully no one would trust them if they are incompetent :)

Anyways, what do you suggest as an alternative?

Reply:
How would a user be able to make an informed decision on how securely the creator of a manifest is able to configure a web server?

I would personally assume that there is no individual that runs a web server in their private capacity that could be able to do it perfectly. So I would personally never use this system as a root of trust for my own sensitive info. Even companies with large, well funded security teams make mistakes which leave web servers temporarily vulnerable to attack. 

I don't have a good idea to solve the key management/distribution issue. For validating manifests, I think that plain signatures are much better.

- Signatures are frozen at a point in time. Every time someone wants to change it they need access to the private key.
- In the web server setting it is stored on disk over time and can change for many reasons without needing access to the private key.
- The attack surface and time window of an attack is hugely expanded in the "integrity of the entire web server as source of trust" scenario.
- There is no validation apart from the transport. If the file became corrrupted on disk the signature would fail to validate but not in the TLS case. If for instance a random bit flip in the storage changes a single byte in one of the hashes in the manifest would stop working, but this would be very difficult to troubleshoot as it would still look like valid JSON.

A transparency log could solve some of the issues by highlighting every time the manifest changes, but that relies on an ecosystem of people monitoring for these changes.

Reply:
Leaving the transparency log aside for now, do you think that a model similar to cargo-crev would be preferable, in which manifests are signed, but assuming that public keys are still discovered and fetched over HTTPS?

Reply:
Yes, I think it is much safer.

Reply:
thanks, changed things quite a bit, I think the resulting design is actually much nicer, PTAL.